Add 0.9.0 release notes (#192)
FEniCS/web: documentation/v0_9_0.md

> ---

Reviewer: Can we drop the raw release notes I created in GitHub Releases to the bottom of this?

Author: We could, but do we really want a copy of them on the webpage?

title: FEniCS release notes
subtitle: v0.9.0
permalink: /blog/v0.9.0/
---

> Since the release of v0.8.0 in May there has been 190 Pull requests merged into DOLFINx from 12 contributors.

Reviewer: pull requests

> Since the release of v0.8.0 in May there has been 190 Pull requests merged into DOLFINx from 12 contributors.

Author: ```suggestion
Since the release of v0.8.0 in May there has been 190 pull requests merged into DOLFINx from 12 contributors.
```

> This release marks a milestone in DOLFINx development, as several new features have been added, see summary below:

Reviewer: Remove see summary below? Seems unnecessary.

> This release marks a milestone in DOLFINx development, as several new features have been added, see summary below:

Author: ```suggestion
This release marks a milestone in DOLFINx development, as several new features have been added.
```

### New features

> - [Mixed-mesh and mixed dimensional assembly](#mixed-assembly)

Author: In the title or the text? Embedded mixed mesh? Conforming mixed mesh?

> - [Mixed-mesh and mixed dimensional assembly](#mixed-assembly)

Author: Co-dimension 0 submesh coupled with parent (or another codim 0 mesh).

- [Simpler constructions of block forms with `ufl.MixedFunctionSpace` and `ufl.extract_blocks`](#extract-blocks)
- [Data-independent form compilation with `dolfinx.fem.compile_form`](#compile-form)
- [Native Windows compatibility](#native-windows)

### API changes

- [Improving `dolfinx.mesh.refine`](#refine)
- [New function for nonmatching interpolation](#nonmatching)
- [Remove `dolfinx.fem.set_bc`](#setbc)
- [Rename `scale` in `apply_lifting` to `alpha`](#lifting)
- [Move PETSc Vector wrapper from `dolfinx.fem.Function.vector` to `dolfinx.fem.Function.x.petsc_vec`](#petsc-mem)

#### Python API additions

- Expose `DirichletBC.dof_indices` to Python API, [DOLFINx \#3389](https://github.com/FEniCS/dolfinx/pull/3389)
- Expose `NewtonSolver.set_convergence_check`, [DOLFINx \#3386](https://github.com/FEniCS/dolfinx/pull/3386)

### Other updates

> - [All users of `PETSc.SNES` should upgrade from PETSc<3.21.6](#snes)

Author: It is not really a bug/feature/fix in DOLFINx but a PETSc issue. So I don’t think it should be moved up.

- [Performance update for form compilation with subdomains](#subdomains)
- [Improved documentation of `dolfinx.mesh`](#mesh)
- [ARM docker images with VTK](#vtk)

- [Initial support for mixed topology meshes](#mixed-mesh)
- [Outputting of tensors of arbitrary size](#tensors)
- [Built-in interpolation matrix](#imatrix)

A detailed summary of all pull requests can be found at [Release: DOLFINx v0.9.0](https://github.com/FEniCS/dolfinx/releases/tag/v0.9.0)

---

# New features

## Mixed dimensional assembly {#mixed-assembly}

**Main contributors:** [Joseph Dean](https://github.com/jpdean/) and [Jørgen S. Dokken](https://github.com/jorgensd/)

> In DOLFINx, we now support assembly of 0th, 1st and 2nd order tensors consisting of meshes from different sub-meshes. One can now create function space over sub-meshes, and assemble with coefficients, test-functions or trial-functions from the parent mesh (and vici versa).  

Reviewer: vici versa -> vice versa

> In DOLFINx, we now support assembly of 0th, 1st and 2nd order tensors consisting of meshes from different sub-meshes. One can now create function space over sub-meshes, and assemble with coefficients, test-functions or trial-functions from the parent mesh (and vici versa).  

Author: ```suggestion
In DOLFINx, we now support assembly of 0th, 1st and 2nd order tensors consisting of meshes from different sub-meshes. 
One can now create function space over sub-meshes, and assemble with coefficients, test-functions or trial-functions from the parent mesh (and vice versa).  
```

> The feature currently works for mixed assembly of co-dimension 0 (sub-set of cells) and co-dimension 1 (sub-set of facets).

Reviewer: sub-set -> subset

> The feature currently works for mixed assembly of co-dimension 0 (sub-set of cells) and co-dimension 1 (sub-set of facets).

Author: ```suggestion
The feature currently works for mixed assembly of co-dimension 0 (subset of cells) and co-dimension 1 (subset of facets).
```

- Codim-1 mixed assembly: [DOLFINx \#3224](https://github.com/FEniCS/dolfinx/pull/3224)
- Interior facet assembly with submesh of codimensional 1: [DOLFINx \#3452](https://github.com/FEniCS/dolfinx/pull/3452)
- Submesh of vertices: [DOLFINx \#3455](https://github.com/FEniCS/dolfinx/pull/3455)
- C++ demo: [DOLFINx \#3436](https://github.com/FEniCS/dolfinx/pull/3436), [DOLFINx \#3262](https://github.com/FEniCS/dolfinx/pull/3262)
- Packing of coefficients for more than 2 sub-meshes: [DOLFINx \#3260](https://github.com/FEniCS/dolfinx/pull/3260), [DOLFINx \#3361](https://github.com/FEniCS/dolfinx/pull/3361)
- Fix for block vector assembly: [DOLFINx \#3346](https://github.com/FEniCS/dolfinx/pull/3346)
- Add create submesh documentation in Python: [DOLFINx \#3112](https://github.com/FEniCS/dolfinx/pull/3112)
- Consistent orientation of sub-entities, [DOLFINx \#3209](https://github.com/FEniCS/dolfinx/pull/3209)
- Interpolation to and from co-dim 0 meshes, [DOLFINx \#3114](https://github.com/FEniCS/dolfinx/pull/3114)

---

## Simpler construction of block forms {#extract-blocks}

**Main contributor:** [Jørgen S. Dokken](https://github.com/jorgensd/)

> Related to the introduction of [mixed dimensional assembly](#mixed-assembly), having an easier way of creating blocked variational forms is crucial for the usability of the feature. This feature can also be used for "standard" block forms, used with `dolfinx.fem.petsc.assemble_*_block/nest`.

Reviewer: This says very little - what is it and what problem does it solve?

Author: It solves the problem that using the blocked interface/nest interface of PETSC required the user to write out the blocks of the matrix (potentially jacobian of residual) and the residual by hand, making the code super verbose, but also error prone when doing differentiation.

as you can see from the example below and the modified demos in DOLFINx it makes it much easier to code up variational forms.

F += ufl.inner(1, q) * ufl.ds

F_blocked = ufl.extract_blocks(F)
F_compiled = dolfinx.fem.form(F_blocked)

# Compute Jacobian on block form
du, dp = ufl.TrialFunctions(W)
J_blocked = ufl.extract_blocks(ufl.derivative(F, u, du) + ufl.derivative(F, p, dp))
J_compiled = dolfinx.fem.form(J_blocked)

# Assemble block matrix and block vector
import dolfinx.fem.petsc
A = dolfinx.fem.petsc.assemble_matrix_block(J_compiled)
A.assemble()
b = dolfinx.fem.petsc.assemble_vector_block(F_compiled, a=J_compiled)

# Assemble in petsc nest vector to use PC-FieldSplit
A_nest = dolfinx.fem.petsc.assemble_matrix_nest(J_compiled)
A_nest.assemble()
b_nest = dolfinx.fem.petsc.assemble_vector_nest(F_compiled)
```

#### Relevant pull requests since last release:

- Simplify DOLFINx demos with `ufl.extract_block`: [DOLFINx \#3450](https://github.com/FEniCS/dolfinx/pull/3450)
- Extract block fixes: [UFL \#308](https://github.com/FEniCS/ufl/pull/308), [UFL \#310](https://github.com/FEniCS/ufl/pull/310), [UFL \#285](https://github.com/FEniCS/ufl/pull/285)

---

## Data-independent form compilation {#compile-form}

Main contributor: [Jørgen S. Dokken](https://github.com/jorgensd/)  

> To unify the C++ and Python interface with UFL and Basix, the user can now create data-independent variational forms for Python (similar to what is done for C++ problems), and then in turn attached data to the compiled form. This means that the same code can be used for both C++ and Python programs.  

Reviewer: Line wrapping.

> To unify the C++ and Python interface with UFL and Basix, the user can now create data-independent variational forms for Python (similar to what is done for C++ problems), and then in turn attached data to the compiled form. This means that the same code can be used for both C++ and Python programs.  

Reviewer: same code - not specific.

> To unify the C++ and Python interface with UFL and Basix, the user can now create data-independent variational forms for Python (similar to what is done for C++ problems), and then in turn attached data to the compiled form. This means that the same code can be used for both C++ and Python programs.  

Author: ```suggestion
To unify the C++ and Python interface with UFL and Basix, the user can now create data-independent variational forms for Python (similar to what is done for C++ problems), and then in turn attached data to the compiled form.
This means that the same code can be used for both C++ and Python programs.  
```

w = ufl.Coefficient(V)
c = ufl.Constant(domain)
e = ufl.Constant(domain)
J = c * e * u * w * ufl.dx(domain=domain)

# Compile form using dolfinx.jit.ffcx_jit
compiled_form = dolfinx.fem.compile_form(
   	MPI.COMM_WORLD, J, form_compiler_options={"scalar_type": dtype}
)
for N in [2, 4, 8, 16]:
    mesh = dolfinx.mesh.create_unit_square(MPI.COMM_WORLD, N, N, dtype=real_type)
    Vh = dolfinx.fem.functionspace(mesh, u.ufl_element())
    uh = dolfinx.fem.Function(Vh, dtype=dtype)
    uh.interpolate(lambda x: x[0])
    wh = dolfinx.fem.Function(Vh, dtype=dtype)
    wh.interpolate(lambda x: x[1])
    eh = dolfinx.fem.Constant(mesh, dtype(3.0))
    ch = dolfinx.fem.Constant(mesh, dtype(2.0))
    # Attach data to create a `dolfinx.fem.Form` that can be used in assembly
    form = dolfinx.fem.create_form(compiled_form, [], mesh, {u: uh, w: wh}, {c: ch, e: eh})
```

#### Relevant pull requests since last release:

- Form independent compilation:[DOLFINx \#3263](https://github.com/FEniCS/dolfinx/pull/3263)
- Form independent compilation with submeshes: [DOLFINx \#3262](https://github.com/FEniCS/dolfinx/pull/3262).

---

## Native Windows compatibility {#native-windows}

> **Main contributors:** [Jack Hale](https://github.com/jhale/) and [Min Ragan-Kelley](https://github.com/minrk/)

Reviewer: Chris helped quite a bit with this.

> **Main contributors:** [Jack Hale](https://github.com/jhale/) and [Min Ragan-Kelley](https://github.com/minrk/)

Author: ```suggestion
**Main contributors:** [Jack Hale](https://github.com/jhale/), [Chris Richardson](https://github.com/chrisrichardson) and [Min Ragan-Kelley](https://github.com/minrk/)
```

> There is now support in DOLFINx for pure windows builds (without PETSc). You can install DOLFINx with conda on Windows, see: [https://github.com/FEniCS/dolfinx/?tab=readme-ov-file\#conda](https://github.com/FEniCS/dolfinx/?tab=readme-ov-file#conda) for details.

Reviewer: windows -> Windows

> There is now support in DOLFINx for pure windows builds (without PETSc). You can install DOLFINx with conda on Windows, see: [https://github.com/FEniCS/dolfinx/?tab=readme-ov-file\#conda](https://github.com/FEniCS/dolfinx/?tab=readme-ov-file#conda) for details.

Reviewer: pure -> native.

> There is now support in DOLFINx for pure windows builds (without PETSc). You can install DOLFINx with conda on Windows, see: [https://github.com/FEniCS/dolfinx/?tab=readme-ov-file\#conda](https://github.com/FEniCS/dolfinx/?tab=readme-ov-file#conda) for details.

Reviewer: would be good to mention this port includes MPI!

> There is now support in DOLFINx for pure windows builds (without PETSc). You can install DOLFINx with conda on Windows, see: [https://github.com/FEniCS/dolfinx/?tab=readme-ov-file\#conda](https://github.com/FEniCS/dolfinx/?tab=readme-ov-file#conda) for details.

Author: ```suggestion
There is now support in DOLFINx for native Windows builds with MPI (without PETSc). You can install DOLFINx with conda on Windows, see: [https://github.com/FEniCS/dolfinx/?tab=readme-ov-file\#conda](https://github.com/FEniCS/dolfinx/?tab=readme-ov-file#conda) for details.
```

- Support graph partitioner in refine: [DOLFINx \#3444](https://github.com/FEniCS/dolfinx/pull/3444)
- Unify refine interface [DOLFINx \#3322](https://github.com/FEniCS/dolfinx/pull/3322)
- Various improvements: [DOLFINx: 3447](https://github.com/FEniCS/dolfinx/pull/3447), [DOLFINx \#3360](https://github.com/FEniCS/dolfinx/pull/3360), [DOLFINx 3151](https://github.com/FEniCS/dolfinx/pull/3151)

---

## Standalone function for non-matching interpolation {#nonmatching}

**Main contributor:** [Jørgen S. Dokken](https://github.com/jorgensd/)

With the introduction of sub-meshes, it became clear that having a single entry-point to `interpolate` for both interpolation from a sub-mesh with a known map, and a non-matching mesh was not possible.  
To fix this, we have decided to create a stand-alone interpolation for non-matching meshes  
`dolfinx.fem.Function.interpolate_nonmatching`.

#### Relevant pull requests since last release:

- Rewrite interpolation API for nonmatching meshes: [DOLFINx \#3177](https://github.com/FEniCS/dolfinx/pull/3177/) for details.

---

## **Remove** `dolfinx.fem.set_bc` {#setbc}

**Main contributor:** [Garth N. Wells](https://github.com/garth-wells)

> The underlying code for setting dirichlet boundary conditions have been simplified, and what was previously done with  

Reviewer: Dirichlet caps

> The underlying code for setting dirichlet boundary conditions have been simplified, and what was previously done with  

Reviewer: have been -> has been

> The underlying code for setting dirichlet boundary conditions have been simplified, and what was previously done with  

Author: ```suggestion
The underlying code for setting Dirichlet boundary conditions have been simplified, and what was previously done with  
```

`dolfinx.fem.set_bc(b.array, bcs)`  
is now done with  
`[bc.set(b.array) for bc in bcs]`  

> **_Note_** that `dolfinx.fem.petsc.set_bc` has **not** changed.

Reviewer: Why?

A note on the memory management of all PETSc objects created with DOLFINx has been added to relevant functions.

#### Relevant pull requests

- Associate PETsc vector with `dolfinx.la.Vector`: [DOLFINx \#3092](https://github.com/FEniCS/dolfinx/pull/3092)
- Note on memory management: [DOLFINx \#3329](https://github.com/FEniCS/dolfinx/pull/3329)
- Remove deprecated method: [DOLFINx \#3435](https://github.com/FEniCS/dolfinx/pull/3435)

---

# Other updates

#### Upgrade from PETSc<3.21.6 {#snes}

A bug was introduced in PETSc 3.21 that means that `newtontr` in `PETSc.SNES` did not work in **serial** in DOLFINx: [PETSc \#1645](https://gitlab.com/petsc/petsc/-/issues/1645).
This has been rectified in PETSc 3.21.6.  
The new **stable** docker images uses PETsc 3.22: [DOLFINx: 3445](https://github.com/FEniCS/dolfinx/pull/3445)

#### Performance improvement for compilation of forms with multiple subdomains {#subdomains}

A fundamental re-design in UFL has been added to reduce the amount of generated integration kernels when using sub-domains, see: [UFL: \#305](https://github.com/FEniCS/ufl/pull/305) for details

#### Documentation improvements for `dolfinx.mesh` {#mesh}

> The `dolfinx.mesh.Topology` and `dolfinx.mesh.Geometry` classes are now fully wrapped in Python and properly documented, [DOLFINx \#3406](https://github.com/FEniCS/dolfinx/pull/3406), [DOLFINx \#3403](https://github.com/FEniCS/dolfinx/pull/3403), [DOLFINx \#3400](https://github.com/FEniCS/dolfinx/pull/3400)

Reviewer: One sentence on what the user could do.

#### ARM VTK support {#vtk}

The [`ghcr.io/fenics/dolfinx/lab`](http://ghcr.io/fenics/dolfinx/lab) images now includes VTK binaries to use with pyvista, thanks to [Henrik Finsberg](https://github.com/finsberg), [DOLFINx \#3259](https://github.com/FEniCS/dolfinx/pull/3259)

#### Create mixed topology mesh {#mixed-mesh}

Initial support by [Chris Richardson](https://github.com/chrisrichardson) for mixed topology meshes, see: [DOLFINx \#3271](https://github.com/FEniCS/dolfinx/pull/3271) for details.  
Relevant PRs: [DOLFINx \#3237](https://github.com/FEniCS/dolfinx/pull/3237), [DOLFINx \#3240](https://github.com/FEniCS/dolfinx/pull/3240), [DOLFINx \#3234](https://github.com/FEniCS/dolfinx/pull/3234), [DOLFINx \#3223](https://github.com/FEniCS/dolfinx/pull/3223)

#### Support arbitrary length tensor in outputting {#tensors}

> Previously only up to 3x3 tensors and length 3 vectors would be outputted. [DOLFINx \#3227](https://github.com/FEniCS/dolfinx/pull/3227) resolves this.

Reviewer: For which formats?

> Previously only up to 3x3 tensors and length 3 vectors would be outputted. [DOLFINx \#3227](https://github.com/FEniCS/dolfinx/pull/3227) resolves this.

Author: ```suggestion
Previously only up to 3x3 tensors and length 3 vectors would be outputted. [DOLFINx \#3227](https://github.com/FEniCS/dolfinx/pull/3227) resolves this for `VTKFile`, `VTXWriter` and `XDMFFile`
```

> #### Interpolation matrix with built in matrices {#imatrix}

Reviewer: built in matrices -> DOLFINx native matrices?

---
title: FEniCS release notes
subtitle: v0.9.0
permalink: /blog/v0.9.0/
---

Since the release of v0.8.0 in May there has been 190 pull requests merged into DOLFINx from 12 contributors.
This release marks a milestone in DOLFINx development, as several new features have been added.

### New features

> - [Mixed-mesh and mixed dimensional assembly](#mixed-assembly)

Author: ```suggestion
- [Mixed dimensional assembly](#mixed-assembly)
```

- [Simpler constructions of block forms with `ufl.MixedFunctionSpace` and `ufl.extract_blocks`](#extract-blocks)
- [Data-independent form compilation with `dolfinx.fem.compile_form`](#compile-form)
- [Native Windows compatibility](#native-windows)

### API changes

- [Improving `dolfinx.mesh.refine`](#refine)
- [New function for nonmatching interpolation](#nonmatching)
- [Remove `dolfinx.fem.set_bc`](#setbc)
- [Rename `scale` in `apply_lifting` to `alpha`](#lifting)
- [Move PETSc Vector wrapper from `dolfinx.fem.Function.vector` to `dolfinx.fem.Function.x.petsc_vec`](#petsc-mem)

#### Python API additions

- Expose `DirichletBC.dof_indices` to Python API, [DOLFINx \#3389](https://github.com/FEniCS/dolfinx/pull/3389)
- Expose `NewtonSolver.set_convergence_check`, [DOLFINx \#3386](https://github.com/FEniCS/dolfinx/pull/3386)

### Other updates

- [All users of `PETSc.SNES` should upgrade from PETSc<3.21.6](#snes)
- [Performance update for form compilation with subdomains](#subdomains)
- [Improved documentation of `dolfinx.mesh`](#mesh)
- [ARM docker images with VTK](#vtk)

- [Initial support for mixed topology meshes](#mixed-mesh)

> - [Outputting of tensors of arbitrary size](#tensors)

Author: ```suggestion
- [Output of tensors of arbitrary size](#tensors)
```
